packages/repository/src/__tests__/unit/repositories/relations-helpers/mongo-related-helpers.unit.ts

```diff
@@ -0,0 +1,202 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/repository-tests
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect, toJSON} from '@loopback/testlab';
+import {ObjectID} from 'bson';
+import {} from 'bson';
+import {testdb} from './relations-helpers-fixtures';
+import {
+  belongsTo,
+  BelongsToAccessor,
+  DefaultCrudRepository,
+  Entity,
+  Getter,
+  hasMany,
+  HasManyRepositoryFactory,
+  juggler,
+  model,
+  property,
+} from '../../../..';
+import {
+  isBsonType,
+  deduplicate,
+  normalizeKey,
+  buildLookupMap,
+  reduceAsArray,
+} from '../../../../relations';
+
+describe('unit tests, simulates mongodb env fßßor helpers of ininclusion resolver ', () => {
+  let productRepo: ProductRepository;
+  let categoryRepo: CategoryRepository;
+
+  before(() => {
+    productRepo = new ProductRepository(testdb);
+    categoryRepo = new CategoryRepository(testdb, async () => productRepo);
+  });
+
+  beforeEach(async () => {
+    await productRepo.deleteAll();
+    await categoryRepo.deleteAll();
+  });
+
+  describe('helpers for formating instances', async () => {
+    it('isBsinType', async () => {
+      const id1 = new ObjectID();
+      const id2 = new ObjectID();
+      expect(isBsonType(id1)).to.be.true();
+      expect(isBsonType(id2)).to.be.true();
+    });
+    it('simple deduplcate + isBsonType', async () => {
+      const id1 = new ObjectID();
+      const id2 = new ObjectID();
+
+      const result = deduplicate([id1, id2]);
+
+      expect(result).to.deepEqual(toJSON([id1, id2]));
+      expect(typeof result[0] === 'string').to.be.true();
+      expect(typeof result[1] === 'string').to.be.true();
+    });
+    it('multiple items deduplcate + isBsonType', async () => {
+      const id1 = new ObjectID();
+      const id2 = new ObjectID();
+      const id3 = new ObjectID();
+      expect(isBsonType(id1)).to.be.true();
+
+      const result = deduplicate([id3, id1, id1, id3, id2]);
+
+      expect(result).to.deepEqual(toJSON([id3, id1, id2]));
+    });
+  });
+
+  describe('helpers for generating inclusion resolvers', async () => {
+    it('normalizeKey', async () => {
+      const id = new ObjectID();
+      expect(normalizeKey(id)).to.eql(id.toHexString());
+    });
+    // the tests below simulate mongodb environment.
+    it('simple normalizeKey + buildLookupMap', async () => {
+      const id = new ObjectID();
+      expect(normalizeKey(id)).to.eql(id.toHexString());
+    });
+    it('normalizeKey + buildLookupMap: returns multiple instances in an array', async () => {
```

Comment:
> `import {testdb} from './relations-helpers-fixtures';`

Are you sharing the same datasource instance between multiple tests? That would be the most likely cause of the behavior your are observing. Juggler does not support redefinition of a model with the same name in the same datasource.

Solutions:
- Rework these tests to unit-test style, don't use any repositories.
- Don't share datasource between tests, create a new one in `before` hook. Just call
    ```ts
    new juggler.DataSource({connector: 'memory'})
    ```
- Call `testdb.deleteAllModels()` before defining any new models (in a `before` hook). This is the least preferred option to me. It has its place in acceptance-level tests in `repository-tests` - we want to reuse the datasource to avoid re-opening database connections, because that would slow down the tests. Everywhere else, where we are testing against `memory` connector only, we should create a new datasource in each test suite.

See also https://github.com/strongloop/loopback-next/pull/3431/commits/38b3f906b891520437b61436d12674d711638a15 and https://github.com/strongloop/loopback-datasource-juggler/pull/1759

Reply:
yes! there is where the problem lies. Thanks!